### PR TITLE
Sharp positions oneoff

### PR DESCRIPTION
**Pull request by sethtroisi**
Script to find positions that differentiate good and bad models.

Takes as input
* directory of SGFs
* directory of models
* ratings of models (wget  cloudygo/<bucket>/ratings.json)

Outputs
* SGF name (and move number)
* coefficients that explain how much elo it attributes to
  * correct value calculation
  * correct value range (AKA "bucket") 0-0.2, 0.2-0.4, 0.4-0.6, 0.6-0.8, 0.8-1.0
  * choosing move X, Y, Z, any other move
  * how sharp policy was for the position

output can be seen at:
https://cloudygo.com/v13-19x19/puzzles/0
and
https://cloudygo.com/v13-19x19/puzzles/0?show

Running this is a multi-step process

```
# Create collection.csv of sgf and move number
SGF_DIR=data/sgf
MODEL_DIR=models/
BOARD_SIZE=19 python3 sharp_positions.py subsample --num_positions 10 --sgf_dir $SGF_DIR
```

```
# Create a directory of those sgfs truncated at the specified move number
# from https://github.com/sethtroisi/go-scripts/
./truncate.sh

# Remove any bad SGFS.
cat badfiles.txt | tqdm | xargs rm

# Rerun truncate successfully this time
./truncate.sh

# Remove original SGFs
cat ../minigo/collection.csv | cut -f1 -d, | sort -u | tqdm | xargs rm
```

```
# Get value & policy for all models for all positions 
python3 sharp_positions.py evaluate --sgf_dir <problem-collections2> --model_dir $MODEL_DIR
```

```
# You can run this multiple times tweaking parameters get a smaller set of positions 
python3 sharp_positions.py minimize \
            --model_dir $MODEL_DIR --sgf_dir data/s
            --rating_json ratings.json --results results.csv
```

### CODE REVIEW

File: oneoffs/heatmap.py [hunk 1]
```diff
@@ -94,8 +94,7 @@ def positions_from_sgfs(sgf_files, include_empty=True):
 
     for sgf in sgf_files:
         sgf_name = os.path.basename(sgf).replace(".sgf", "")
-        sgf_positions, moves, _ = oneoff_utils.parse_sgf(sgf)
-        final = sgf_positions[-1].play_move(moves[-1])
+        final = oneoff_utils.final_position_sgf(sgf)
```
Comment (amj) [inline oneoffs/heatmap.py]: should these be rebased?  this looks like stale changes.

File: oneoffs/sharp_positions.py [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+flags.DEFINE_integer("num_positions", 1, "How many positions from each game.")
+flags.DEFINE_integer("top_n", 3, "Policy moves to record per position.")
+flags.DEFINE_integer("min_idx", 100, "Min model number to include.")
+
+flags.DEFINE_string("sgf_dir", "data/s/", "collection of SGFs")
+flags.DEFINE_string("model_dir", "models/", "Folder of models")
+flags.DEFINE_string("rating_json", "ratings.json", "ratings of models")
+flags.DEFINE_string("results", "results.csv", "Evaluate results file")
+flags.DEFINE_string("SVM_json", "SVM_data.json", "SVM data about positions")
```
Comment (amj) [inline oneoffs/sharp_positions.py]: your docstrings here should make clear the difference between inputs and outputs, even if its just from tenses.  "Evaluate results file" is ambiguous

File: oneoffs/sharp_positions.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+
+def grouper(n, iterable):
+    """Itertools recipe
+    >>> list(grouper(3, iter('ABCDEFG')))
+    [['A', 'B', 'C'], ['D', 'E', 'F'], ['G']]
+    """
+    return (iterable[i:i + n] for i in range(0, len(iterable), n))
+
+
+def pretty(v):
```
Comment (amj) [inline oneoffs/sharp_positions.py]: this looks off.

File: oneoffs/sharp_positions.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+    return int(number)
+
+
+def subsample():
+    """Sample num_positions postions from each game in sgf_dir
+
+    Usage:
+        python3 sharp_positions.py subsample --num_positions 10 --sgf_dir data/s
+
+    NOTE(sethtroisi): I wrote a truncate script to turn these into SGFS see
```
Comment (amj) [inline oneoffs/sharp_positions.py]: to turn *what* into sgfs?

File: oneoffs/sharp_positions.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+        https://github.com/sethtroisi/go-scripts
+    """
+
+    sgf_files = oneoff_utils.find_and_filter_sgf_files(
+        FLAGS.sgf_dir, None, None)
+
+    random.seed(1)
```
Comment (amj) [inline oneoffs/sharp_positions.py]: flag?

File: oneoffs/sharp_positions.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+    ratings = json.load(open(FLAGS.rating_json))
+    raw_ratings = {int(r[0]): float(r[1]) for r in ratings}
+
+    model_ratings = []
+    for model in model_paths:
+        model_idx = get_model_idx(model)
+        if model_idx < FLAGS.min_idx:
+            continue
+
+        # NOTE: Early models aren't present please skip them with FLAGS.min_idx
```
Comment (amj) [inline oneoffs/sharp_positions.py]: nit: 'If early'...?

File: oneoffs/sharp_positions.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+    labels_train = model_ratings[train_models]
+    labels_test = model_ratings[test_models]
+
+    # Choose some set of positions and see how well they explain ratings
+    positions_to_use = set(positions.keys())
+    linearSVM = svm.LinearSVR()
+    best_test_r2 = 0
+    below_threshold = 0
+
+    for iteration in itertools.count(1):
```
Comment (amj) [inline oneoffs/sharp_positions.py]: just one iteration?

File: oneoffs/sharp_positions.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+        # Drop any positions that aren't very useful
+        for dropped, pos_idx in enumerate(pos_value_idx[:filter_limit], 1):
+            contribution = position_coefs[pos_idx]
+            positions_to_use.remove(iter_positions[pos_idx])
+            print("\t\tdropping({}): {:.1f} {}".format(
+                dropped, contribution, sgf_names[iter_positions[pos_idx]]))
+
+            if contribution > filter_contribution_percent * most_value:
+                break
+        print()
```
Comment (amj) [inline oneoffs/sharp_positions.py]: ?

File: oneoffs/sharp_positions.py [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+
+        # Break at end with
```
Comment (amj) [inline oneoffs/sharp_positions.py]: unfinished sentence?

File: oneoffs/sharp_positions.py [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+
+    Usage:
+        python3 sharp_positions.py evaluate --sgf_dir data/s --model_dir models/
+    """
+
+    # Load positons
+    sgf_names, all_positions = get_final_positions()
+
+    # Run and save some data about each position
+    # Save to csv because that's easy
```
Comment (amj) [inline oneoffs/sharp_positions.py]: csv w/o a csv-writer is asking for trouble with escaping :)  Why not json?

File: oneoffs/sharp_positions.py [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+                tf.reset_default_graph()
+                player = None
+
+            if player:
+                oneoff_utils.restore_params(model, player)
+            else:
+                player = oneoff_utils.load_player(model)
+
+            row = [model]
+            for group, positions in enumerate(grouper(64, all_positions)):
```
Comment (amj) [inline oneoffs/sharp_positions.py]: 64 == batch size?

File: oneoffs/sharp_positions.py (hunk 7 above)
Comment (sethtroisi) [reply oneoffs/sharp_positions.py]: itertools.count() is a infinite loop, the 1 just indicates where to start.

File: oneoffs/sharp_positions.py (hunk 8 above)
Comment (sethtroisi) [reply oneoffs/sharp_positions.py]: This makes the output much nicer.

```
iter 404, 300/2728 included, R^2: 0.9736 train, 0.966 test
	Most value 98 => 44.6 KongJie-LiKang52637-243.sgf
		dropping(1): 13.4 YangYi-KimSeungjun33403-279.sgf

iter 405, 300/2727 included, R^2: 0.9723 train, 0.961 test
	Most value 16 => 46.5 ChenYaoye-ZhongWenjing39960-32.sgf
		dropping(1): 14.1 QiuJun-ShaoWeigang54688-161.sgf

iter 406, 300/2726 included, R^2: 0.9641 train, 0.960 test
	Most value 85 => 45.2 KeJie-QinYuexin65537-40.sgf
		dropping(1): 3.8 LinFeng-HuangYizhong31978-142.sgf
		dropping(2): 11.1 ZhouHeyang-WangLei35439-263.sgf
		dropping(3): 12.8 DongMengcheng-QiuJun66053-197.sgf
		dropping(4): 13.9 LiKang-LeeYounggu41412-103.sgf

iter 407, 300/2722 included, R^2: 0.9692 train, 0.962 test
	Most value 23 => 50.0 ChoiCheolhan-YangDingxin66034-56.sgf
		dropping(1): 10.5 WangHaoyang-ZhongWenjing41815-182.sgf
		dropping(2): 11.4 ZhengYan-LiuXing35540-188.sgf
		dropping(3): 12.7 WangXi-AnDongxu46608-322.sgf
		dropping(4): 13.7 ChunGuangya-YangDingxin51999-222.sgf
		dropping(5): 13.8 YuBin-ChenYaoye34465-131.sgf
		dropping(6): 14.7 MaYichao-HuangYunsong53476-117.sgf
		dropping(7): 15.3 ZhouHeyang-PengQuan40173-97.sgf
```

**Pull request merged**